File: Python/Nonlineq/Bisection.py

import numpy as np #Basic linear algebra
import matplotlib.pyplot as plt #Plotting


tol = 10*10**(-15)
def f(x):
    return x**3-3*x-5
def f_prime(x):
    return 3*x**2-3
#%%
#Halveringsmetoden

n=0 #intervall
p=5
halvering_x=[]
halvering=[]
k=0
while abs(n-p)>tol:
    m = (n+p)/2
    if f(m)<0:
        n=m
    else:
        p=m
    halvering_x.append(p)
    halvering.append(f(p))
    k += 1

i=np.linspace(1,k,k) #stara från 1 iteration
plt.plot(i,halvering,"r.-",label="Halveringsmetoden")
plt.legend()
#%%
x0 = 2
x1 = 1
sekant_x=[]
sekant=[]
sekant_x.append(x0);sekant_x.append(x1)
sekant.append(f(x0));sekant.append(f(x1))
k2 = 2 #gissningar räknas som iteration
while abs(x1-x0)>tol:
    x = x1-f(x1)*((x0-x1)/(f(x0)-f(x1)))
    x0 = x1
    x1 = x
    sekant_x.append(x)
    sekant.append(f(x)) #spara f(x) i array för senare plot
    k2 += 1
i2=np.linspace(1,k2,k2) #stara från 1 iteration
plt.plot(i2,sekant,"bo-",label="Sekantmetoden") #plotta antal iteration och f(x)
plt.legend()


plt.xlabel('iteration')
plt.ylabel('funktion x**3-3*x-5')

plt.title('Sekant ocb halvering på [0,5] för x(1)=' f'{sekant_x[0]}' ',x(2)=' f'{sekant_x[1]}')

File: MATLAB/Nonlineq/Bisection.m
close all;

tolerance = 10e-15;

fun = @(x)x^3 - 3*x -5;

% definition av intervalet [n,p]

n= 0;
p=3;
it=0;

% Halveringsmetoden

	  while abs(n - p) > tolerance
m = (n + p)/2;
         if fun(m) < 0 
  n = m;
         else
	   p = m;
        end

	it = it+1;

func_val(it) =fun(p);
iteration(it) = it;
end



%Sekantmetoden

x=0;
iteration_sekant = 0;
func_sekant = 0;
k=2;
%Initialiseringen
x(1) = 2.0;
x(2) =  1.0;

iteration_sekant(1) = 1;
iteration_sekant(2) = 2;
func_sekant(1) =fun(x(1));
func_sekant(2) =fun(x(2));


while abs( x(k) - x(k-1) ) > tolerance

numerator =  fun(x(k))*(x(k-1) - x(k));
denominator = fun(x(k-1)) - fun(x(k));

x(k+1) = x(k) - numerator/denominator;

k = k+1;
iteration_sekant(k) = k;
func_sekant(k) =fun(x(k));


end


	
figure

plot(iteration,  func_val,'o r-',  'LineWidth',2)

  hold on
plot(iteration_sekant,func_sekant, 'o b-',  'LineWidth',2)
  
  xlabel('iteration')
  ylabel('funktion x^3 - 3x -5')
  legend('Halveringsmetoden','Sekantmetoden')
  title(['Sekant och halvering  på [0,3]  för x(1)=',num2str(x(1)),', x(2)= ',num2str(x(2))])


  iteration
  
iteration_sekant

  p

  x
